src/models/karaoke.py

```diff
@@ -0,0 +1,220 @@
+
+    def calculate_char_fade_offsets(self, style: Style, transitionDuration: timedelta):
+        fontScaler = FontScaler(style.fontName, style.fontSize)
+
+        lineCharTimes = [
+            timedelta(milliseconds=m)
+            for m in fontScaler.split_by_rendered_width(
+                pyass.timedelta(transitionDuration).total_milliseconds(), self.text
+            )
+        ]
+
+        charAccTime = timedelta()
+        for char, lineCharTime in zip(self.chars, lineCharTimes):
+            char.fadeOffset = charAccTime
+            charAccTime += lineCharTime
+
+
+def preproc_line_text(line: SongLine, lineNum: int = 0) -> KLine:
+    timedeltaUpToIdx = reduce(
+        lambda a, b: a + [a[-1] + b.length], line.syllables, [timedelta(0)]
+    )
+
+    kLine = KLine(
+        start=line.start,
+        end=line.end,
+        duration=line.end - line.start,
+        kara=[],
+        text=line.romaji,
+        startActor=line.actors[0],
+        actorSwitches=[
+            (timedeltaUpToIdx[breakpoint], actor)
+            for breakpoint, actor in zip(line.breakpoints, line.actors)
+            if breakpoint != 0
+        ],
+        isSecondary=line.isSecondary,
+        isAlone=line.romaji == line.en,
+        isEN=False,
+        lineNum=lineNum,
+    )
+
+    accLength = timedelta()
+    totalChars = 0
+    counts = [a - b for a, b in zip(line.breakpoints[1:], line.breakpoints[:-1])] + [
+        len(line.syllables) - line.breakpoints[-1]
+    ]
+    actors = [
+        actor for (count, actor) in zip(counts, line.actors) for _ in range(count)
+    ]
+
+    for syl, actor in zip(line.syllables, actors):
+        kSyl = KSyl(
+            start=accLength,
+            duration=syl.length,
+            end=accLength + syl.length,
+            text=syl.text,
+            inlineFx=actor,
+            chars=[],
+            i=len(kLine.kara),
+            line=kLine,
+        )
+
+        for c in syl.text:
+            kChar = KChar(
+                char=c,
+                i=totalChars,
+                sylI=len(kSyl.chars),
+                syl=kSyl,
+                line=kLine,
+            )
+
+            kSyl.chars.append(kChar)
+            totalChars += 1
+
+        kLine.kara.append(kSyl)
+
+        accLength += syl.length
+
+    return kLine
+
+
+def preproc_line_text_en(line: SongLine, lineNum: int = 0) -> KLine:
+    timedeltaUpToIdx = reduce(
+        lambda a, b: a + [a[-1] + b.length], line.syllables, [timedelta(0)]
+    )
+    kLineEN = KLine(
+        start=line.start,
+        end=line.end,
+        duration=line.end - line.start,
+        kara=[],
+        text=line.en,
+        startActor=line.actors[0],
+        actorSwitches=[
+            (timedeltaUpToIdx[breakpoint], actor)
+            for breakpoint, actor in zip(line.breakpoints, line.actors)
+            if breakpoint != 0
+        ],
+        isSecondary=line.isSecondary,
+        isAlone=line.romaji == line.en,
+        isEN=True,
```

Comment:
I feel like it'd be cleaner if we could somehow find a way to encapsulate the idea of this variable in a way that's divorced from "EN" since that's a Song-ish concept that doesn't really have much meaning in the K-effect world

Any suggestions on a different name?

Reply:
Right now, my goal for this isn't for it to be a complete kfx generator replacement (like what karaoke templates are), but more of an improvement for the current `.ass` generator to remove the karaoke templater step whenever possible (because debugging karaoke templates are a pain). Since every song in the current database has the concept of Romaji/EN (and secondary/alone by extension), I think it's fine to keep the EN flag in it as-is.

A complete karaoke templater replacement would belong more in `pyass`, and would require an almost 1:1 port of `karaskel.lua`.